### ISSUE TITLE
Generalize wiki#toc#gather_entries

### ISSUE THREAD

**Issue by lervag** [lervag/wiki.vim#288]
`wiki#toc#gather_entries` assumes ATX headers:

https://github.com/lervag/wiki.vim/blob/0a16c4a6649e101d2b4e708665c6628f27fb0735/autoload/wiki/toc.vim#L153-L156

wiki.vim is meant to be filetype agostic, thus this should be generalized to allow other formats, as suggested by @saccarosium in #287.

Proposal: Add an option `g:wiki_toc_header_format` to allow other header formats, e.g.:

```
*g:wiki_toc_header_format*
  Option to specify the header format used for parsing the table of contents.

  Possible values:

    atx ~
      Atx headers are probably the most common header type in Markdown files
      and look like this: >markdown

        # Header level 1

        ## Header level 2
<
    setext ~
      Setext headers are common in Markdown files and look like this: >markdown

        Header level 1
        ===

        Header level 2
        ---
<
    asciidoc ~
      AsciiDoc headers are used in AsciiDoc files. They look like this: >asciidoc

        = Document Title (level 0)

        == Level 1 Section title

        === Level 2 Section title
<
```

One problem with using a global option is that one may want wiki.vim to work with different filetypes. Thus, it should be possible to also override the global option in any relavant commands, e.g. `WikiTocGenerate`:

```
*<plug>(wiki-toc-generate)*
*WikiTocGenerate*
`:[range]WikiTocGenerate [header_format]
  Create/Update table of contents. The optional argument `header_format` allows
  to override the header format specified in |g:wiki_toc_header_format|.
```

**Comment by lervag** [comment]
@saccarosium What do you think; would the above proposal work for you?

**Comment by saccarosium** [comment]
I was thinking more of an automatic behaviour based on the main filetype, the first in the `g.wiki_filetypes` table/dictionary. Contrary to links I think that headings are more "save" to make automatic guess. For example asciidoc heading will be always `=` or in orgmode `*`, is really necessary specify that I'm using asciidoc style heading if I choose asciidoc? 

Having said that I think that for markdown is mandatory due to the infinity flavours and ways to do heading.

Could make the variable an set it automatically to the filetype.

**Comment by lervag** [comment]
Yes, good point. I think my proposal can still hold, except with the following addendum:

```
*g:wiki_toc_header_format*
  Option to specify the header format used for parsing the table of contents.
  This is a dictionary that maps a filetype to a header format. The supported
  header formats are:

    ... (list of supported formats)

  Default value: >vim

    let g:wiki_toc_header_format = {
          \ 'md': 'atx',
          \ 'asciidoc': 'asciidoc',
          \ 'org': 'org',
          \}
<
```

What do you think? This seems to be quite flexible to me, and if we still allow to specify the header format with `:WikiTocGenerate FORMAT`, then I would think most people would be happy?

**Comment by saccarosium** [comment]
> What do you think? This seems to be quite flexible to me, and if we still allow to specify the header format with :WikiTocGenerate FORMAT, then I would think most people would be happy?

Yes indeed I think is looking very good.

**Comment by lervag** [comment]
Then I'll start working on this as soon as I get the time.

**Comment by lervag** [comment]
This is fixed by @saccarosium in #295. Thanks!
